daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala

```diff
@@ -4,73 +4,142 @@
+      final def toError: Interpretation = Interpretation(this)
+      override def detailMsg: String = msg
+    }
+
+    // TODO: get ride of Generic
+    final case class Generic(override val msg: String, override val detailMsg: String)
+        extends SubError
+
+    object Generic {
+      def apply(msg: String): Generic = new Generic(msg, msg)
+    }
+
+    final case class ContractNotFound(ci: Value.ContractId) extends SubError {
+      override def msg = s"Contract could not be found with id $ci"
+    }
+
+    final case class GlobalKeyNotFound(globalKey: GlobalKeyWithMaintainers) extends SubError {
+      override def msg = s"dependency error: couldn't find key ${globalKey.globalKey}"
+    }
+
+    /** See com.daml.lf.transaction.Transaction.DuplicateContractKey
+      * for more information.
+      */
+    final case class DuplicateContractKey(key: GlobalKey) extends SubError {
+      override def msg = s"Duplicate contract key $key"
+    }
+
+    final case class Authorization(override val msg: String) extends SubError {
+      override def detailMsg: String = msg
+    }
+
   }
 
-}
+  // Error happening during transaction validation
+  final case class Validation(subError: Validation.SubError) extends Error
 
-final case class ContractNotFound(ci: ContractId) extends Error {
-  override def msg = s"Contract could not be found with id $ci"
-  override def detailMsg: String = msg
-}
+  object Validation {
 
-/** See com.daml.lf.transaction.Transaction.DuplicateContractKey
-  * for more information.
-  */
-final case class DuplicateContractKey(key: GlobalKey) extends Error {
-  override def msg = s"Duplicate contract key $key"
-  override def detailMsg: String = msg
-}
+    sealed abstract class SubError extends Error.SubError {
+      final def toError: Validation = Validation(this)
+      override def detailMsg: String = msg
+    }
 
-final case class ValidationError(override val msg: String)
-    extends RuntimeException(s"ValidationError: $msg", null, true, false)
-    with Error {
-  override def detailMsg: String = msg
-}
+    // TODO: get ride of Generic
+    final case class Generic(msg: String, override val detailMsg: String) extends SubError
+    object Generic {
+      def apply(msg: String): Generic = new Generic(msg, msg)
+    }
 
-final case class ReplayMismatch(
-    mismatch: transaction.ReplayMismatch[transaction.NodeId, ContractId]
-) extends RuntimeException(s"ValidationError: ${mismatch.msg}", null, true, false)
-    with Error {
-  override def msg: String = mismatch.msg
-  override def detailMsg: String = mismatch.msg
-}
+    final case class ReplayMismatch(
+        mismatch: transaction.ReplayMismatch[transaction.NodeId, Value.ContractId]
+    ) extends SubError {
+      override def msg: String = mismatch.msg
 
-final case class AuthorizationError(override val msg: String) extends Error {
-  override def detailMsg: String = msg
-}
+      override def detailMsg: String = ""
```

Comment:
do we really want the detail to be empty? I thought the idea is that it is either identical to `msg` or contains more info.

Reply:
fixed.